### DISCUSSION TITLE
Problem to plot with pyvista and ipyvtklink

### DISCUSSION THREAD

**Discussion by jc-barreto** [pyvista/pyvista#2040] [Q&A]
Hello,
I'm trying to plot with pyvista and I was getting an static image with an error

![image](https://user-images.githubusercontent.com/42352209/150136217-0ac770fd-25a6-49d3-b6d1-ce4958648605.png)

So, as the warning indicates, I installed ipyvtklink through !pip install ipyvtklink (I'm using Jupyter Notebook).

And now I have no image at all!

What I should do?

**Comment by DanielJamesEvans** [comment]
I had this issue.  Installing an older version of the `ipywidgets` library solved the problem.  (`pip install ipywidgets==7.7.1` was the command.)  Credit to jorgensd at https://github.com/pyvista/pyvista/issues/3274 for finding the workaround.

**Reply by banesullivan** [reply]
I recommend this ^

**Reply by banesullivan** [reply]
Also, sometimes you may need to restart jupyter lab to let `ipyvtklink`'s jupyter widget dependencies to get properly registered

**Reply by astroboylrx** [reply]
@banesullivan Sorry if I missed some documentation. Are some backends of `pyvista` supposed to only work with Jupyter Lab? I downgraded `ipywidgets` and am able to generate interactive plots in Jupyter Lab. However, in Jupyter Notebook, the backend `pythreejs` still produces nothing at all.

**Reply by banesullivan** [reply]
> Are some backends of pyvista supposed to only work with Jupyter Lab?

I'm actually not totally sure. Jupyter Lab goes a long way in managing widgets so that they are easier to install and use, while legacy Jupyter Notebook often has a lot of issues with modern widgets (in my experience). For this reason, I'd be inclined to move towards only support Lab, but I think all the widgets should work fine in Notebook as long as their dependencies are all properly installed and registered with Jupyter.

> However, in Jupyter Notebook, the backend pythreejs still produces nothing at all.

I'm not surprised... @akaszynski, might have some insightful suggestions to get this working

**Reply by DanielJamesEvans** [reply]
With the latest version of pythreejs, I don't get any output in Notebook.  With pythreejs==2.1.0, 2.0.0, or 1.1.0, I get a blank plot.  (This is different from getting no plot at all.)

@astroboylrx Do you need pythreejs, or would another plotting backend be acceptable?  `ipygany` and `ipyvtklink` work for me.

@banesullivan regarding Jupyter Notebook support: I'm not a pyvista developer; please do whatever seems best.  And I certainly appreciate your making this project available, regardless of what you decide!  But I figured I'd cast a vote for keeping Jupyter Notebook support for now... I'm currently writing a package for analyzing biomolecules; AFAIK many people in my field still haven't heard of Jupyter Lab (even if they use Notebooks).  They'll probably have to switch eventually, but the transition might be slow.

**Reply by banesullivan** [reply]
re Lab vs Notebook: I hope we can continue to support both moving forward as the goal is to "make 3D viz approachable" - missing support for Notebook wouldn't be approachable IMO. But it's great to capture this from you as I am currently working on some Jupyter plotting improvements and was hoping I just sideline legacy Notebook in favor of Lab for simplicity's sake... but perhaps we need to better support Notebook *and* Lab

**Reply by astroboylrx** [reply]
@banesullivan Thanks a lot for the reply! I see. I understand that Lab would likely have more prospect than Notebook.
@DanielJamesEvans Many thanks for the info about other versions of `pythreejs`. Yes, other backends works. I asked b/c `ipyvtklink` render on server so it is often a bit blur and laggy, and `ipygany` hasn't been updated for 2 years, which seems to make `pythreejs` a more promising choice for long-term use.

**Reply by banesullivan** [reply]
FYI, I am working on a new notebook rendering solution that I hope will be the default and go to for the long run... stay tuned. Mentioning so that you don't get too bought in to one of thee external widgets (that we do not maintain)

**Reply by DanielJamesEvans** [reply]
Thanks for mentioning the upcoming changes to Notebook rendering.  Will the new render engine support `link_views()`?  This command is why I use PyVista; I couldn't find this capability in other libraries (e.g. trimesh).  iirc I couldn't get `link_views` to work with some rendering backends, but that was a while ago.

OTOH I have no data for how many other people use `link_views`.  So if it's super niche and would be annoying to implement with the new engine, I'd understand if it doesn't get included.  If that's the case, I could happily keep using `ipyvtklink` for my project.

**Reply by banesullivan** [reply]
> Will the new render engine support link_views()

Most definitely. FWIW, `ipyvtklink` notebook rendering supports this... which after typing this I realize that's what you're currently using

> OTOH I have no data for how many other people use link_views. So if it's super niche

It's a widely used feature and not niche at all! It's a core part of the plotting API at this point

**Reply by darikg** [reply]
> FYI, I am working on a new notebook rendering solution that I hope will be the default and go to for the long run... stay tuned. Mentioning so that you don't get too bought in to one of thee external widgets (that we do not maintain)

that's exciting! happy to help in any way

**Comment by banesullivan** [comment]
@jc-barreto, could you share a `Report()`? 

```py
import pyvista as pv
print(pv.Report())
```

**Reply by jc-barreto** [reply]
Hello! I solved the problem, I don't remember exactly how, I think it was an incompatibility between versions, so I re-install everything paying attention to the versions, and now works =)

**Comment by Keven-Gao** [comment]
I suffer the same problem, simply install `ipywidgets` and restart jupyter notebook could solve it.
